notebooks/USE_embeddings.py

```diff
@@ -0,0 +1,220 @@
+                extracted_paragraphs.append(cleaned_text_segment)
+            original_text = original_text[end_idx + 3:]
+    return extracted_paragraphs
+
+
+def document_embedding(paragraphs):
+    '''
+    average embeddings across sentences
+    '''
+    embedding = embed(paragraphs)
+    average_embedding = tf.math.reduce_mean(embedding, axis=0).numpy()
+    return average_embedding
+
+
+# initialise an empty array for embeddings
+collected_doc_embeddings = np.zeros((subset_content_df.shape[0], 512))
+
+# fill array with embeddings for all docs
+# problem with out of memory warnings on 8gb card: reduce extracted paragraphs to 128
+# flake8 complaining with the complexity of code beneath which generates embeddings using the description
+'''
+for i in range(subset_content_df.shape[0]):
+    try:
+        doc = subset_content_df.iloc[i]['details']['body']
+        if type(doc) == list:
+            doc = doc[0]['content']
+    except KeyError:
+        try:
+            doc = subset_content_df.iloc[i]['description']
+            if isNaN(doc):
+                continue
+        except KeyError:
+            continue
+    extracted_paragraphs = extract_paragraphs(doc)
+    if len(extracted_paragraphs) > 128:
+        extracted_paragraphs = extracted_paragraphs[:128]
+    if len(extracted_paragraphs) > 0:
+        doc_embedding = document_embedding(extracted_paragraphs)
+        collected_doc_embeddings[i, :] = doc_embedding
+    if i % 1000 == 0:
+        progress = i / subset_content_df.shape[0]
+        print('%s' % float('%.2g' % progress))
+
+# converts embeddings array into dataframe, with content id as unique key
+embeddings_df = pd.DataFrame(collected_doc_embeddings)
+embeddings_df['content_id'] = subset_content_df['content_id'].to_list()
+
+# initialise list for storing raw text
+collected_doc_text = []
+
+# store the original raw text extracted from the documents
+for i in range(subset_content_df.shape[0]):
+    try:
+        doc = subset_content_df.iloc[i]['details']['body']
+        if type(doc) == list:
+            doc = doc[0]['content']
+    except KeyError:
+        try:
+            doc = subset_content_df.iloc[i]['description']
+            if isNaN(doc):
+                collected_doc_text.append('')
+                continue
+        except KeyError:
+            collected_doc_text.append('')
+            continue
+    extracted_paragraphs = extract_paragraphs(doc)
+    if len(extracted_paragraphs) > 0:
+        collected_doc_text.append(' '.join(extracted_paragraphs))
+    else:
+        collected_doc_text.append('')
+    if i % 1000 == 0:
+        progress = i / subset_content_df.shape[0]
+        print('%s' % float('%.2g' % progress))'''
+
+# fill array with embeddings for all docs
+for i in range(subset_content_df.shape[0]):
+    try:
+        doc = subset_content_df.iloc[i]['details']['body']
+    except KeyError:
+        continue
+    extracted_paragraphs = extract_paragraphs(doc)
```

Comment:
Question here about whether this is all that necessary given we are trying to clean the text and extract paragraphs. Surely the description or text columns are what we want?